src/utils/httpRequest.ts

```diff
@@ -4,82 +4,145 @@
  *--------------------------------------------------------------------------------------------*/
 
 import * as fse from 'fs-extra';
-import * as https from 'https';
-import request = require('request');
-import url = require('url');
-import { addUserAgent } from './addUserAgent';
-
-function convertToOptions(options: https.RequestOptions | string): https.RequestOptions {
-    if (typeof options === 'string') {
-        // Must use Node's url, not vscode.Uri
-        let optionsAsUrl = url.parse(options);
-        return <https.RequestOptions>optionsAsUrl;
+import { default as fetch, Request, RequestInit, Response } from 'node-fetch';
+import { URL, URLSearchParams } from 'url';
+import { localize } from '../localize';
+
+export class HttpResponse2<T> {
+    private bodyPromise: Promise<T> | undefined;
+    private normalizedHeaders: { [key: string]: string } | undefined;
+
+    public constructor(private readonly innerResponse: Response) { }
+
+    public async json(): Promise<T> {
+        if (!this.bodyPromise) {
+            // This allows multiple calls to `json()` without eating up the stream
+            this.bodyPromise = this.innerResponse.json();
+        }
+
+        return this.bodyPromise;
+    }
+
+    public get headers(): { [key: string]: string } {
+        if (!this.normalizedHeaders) {
+            this.normalizedHeaders = {};
+            for (const key of this.innerResponse.headers.keys()) {
+                this.normalizedHeaders[key] = this.innerResponse.headers.get(key);
+            }
+        }
+
+        return this.normalizedHeaders;
+    }
+}
+
+export function basicAuthHeader(username: string, password: string): string {
+    const buffer = Buffer.from(`${username}:${password}`);
+    return `Basic ${buffer.toString('base64')}`;
+}
+
+export function bearerAuthHeader(token: string): string {
+    return `Bearer ${token}`;
+}
+
+export async function httpRequest2<T>(url: string, options?: RequestOptionsLike, signRequest?: (request: RequestLike) => Promise<RequestLike>): Promise<HttpResponse2<T>> {
```

Comment:
Is the `2` in `httpRequest2` still necessary?

Reply:
🤦‍♂️ I knew I forgot something!

I had it that way so I could easily refactor and change the name later (because it was distinct from any other symbol in the code base)